### PR TITLE
[WIP] add option to convert bbcode to markdown

### PR DESCRIPTION
**Pull request by Dahie**
This is still work in progress, but shows a possible direction discussed in #22 

The tests are failing at the moment. I have a bit of a hard time with Treetop, it's documentation is … sparse with a steep learning curve.

### CODE REVIEW

File: test/doc_parsing_test.rb [hunk 1]
```diff
@@ -1,61 +1,103 @@
...
+
+A paragraph with a line break.
+
+Here\'s one kind of link: [http://example.org](http://example.org)
+
+Here\'s another: [Example](http://example.com)
+
+This is an image: ![Image](http://example.com/foo.png)
+
+>  This *is* a quotation.
```
Comment (Dahie) [inline test/doc_parsing_test.rb:75]: The current treetop node structure does not allow to have these quotes per line. I need some help here.

Comment (jarrett) [reply test/doc_parsing_test.rb:75]: Sorry for the delay. Do you still need help with this? If so, could you elaborate on what you're trying to do and what problems you ran into?

**Comment by Dahie** [comment]
Yes, still open. So the goal is to translate

    [quote]this is
    
    a 
    multiline quote
    [/quote]
    

into 

    > this is
    >
    > a
    > multine quote
    >
    

My thinking went in to directions to achieve this. Either by replacing the newlines within the block node to `\n> ` but that messes with included paragraphs and leaves empty new lines so far.
Thinking two was having this dealt properly in Parslet and introducing `QuoteLineItems`or something that get prepended with the `> `

**Comment by jarrett** [comment]
Ah, thanks for pointing that out. So tell me if I'm understanding this correctly:

In BBCode, the string `apples\noranges` is supposed to parse to the same HTML as `apples oranges`. In other words, the single newline character is treated as a space. But when converting to Markdown, you actually do want to preserve the single newline character in the output.

Did I get that right? If so, then I think your second option is the way to go. I.e. we'll have to extend the `.treetop` file so it emits different nodes for `apples\noranges` and `apples oranges`. Then, we'll add logic to consume these types of nodes in the HTML and Markdown rendering methods.

**Comment by jarrett** [comment]
BTW, I added comments to `node_extensions.rb` in the master branch just now. Hopefully this makes the interplay between the `.treetop` file and `node_extensions.rb` easier to understand.

**Comment by Dahie** [comment]
> Ah, thanks for pointing that out. So tell me if I'm understanding this correctly:
> 
> In BBCode, the string `apples\noranges` is supposed to parse to the same HTML as `apples oranges`. In other words, the single newline character is treated as a space. But when converting to Markdown, you actually do want to preserve the single newline character in the output.
> 
> Did I get that right? If so, then I think your second option is the way to go. I.e. we'll have to extend the `.treetop` file so it emits different nodes for `apples\noranges` and `apples oranges`. Then, we'll add logic to consume these types of nodes in the HTML and Markdown rendering methods.

Good point, markdown's handling of when to paragraph and when to line breaks got me a few times already. It's documented here: https://www.markdownguide.org/basic-syntax/#paragraphs-1

    [quote]this becomes    
    a 
    single line quote
    [/quote]
    

would actually become

    > this becomes a single line quote
    

While 

    [quote] this becomes
      
    a  
    multiline quote
    [/quote]
    

should become

    > this becomes
    >
    > a
    > multiline quote
    
      
(Note the double-spaces behind the "a")

**Comment by Dahie** [comment]
Btw, what is the minimum supported ruby version of the gem? I'd tend to go for [2.5 given everything older is EOL](https://www.ruby-lang.org/en/downloads/)

File: test/doc_parsing_test.rb [hunk 2]
```diff
@@ -1,61 +1,104 @@
...
-        Smile: :) Frown: :(
-      ',
+        <p>Smile: <img src="/happy.png" alt="Emoticon"/> Frown: <img src="/sad.png" alt="Emoticon"/></p>
+
+    EOS
+  end
+
+  def expected_markdown
+    <<-EOS
+**This *is* awesome**
```
Comment (Dahie) [inline test/doc_parsing_test.rb:64]: This fails at the moment and creates the line `**This [i]is[/i] awesome**`
Something in the recursive detection seems to be broken.

Comment (jarrett) [reply test/doc_parsing_test.rb:64]: Thanks! I think this should be doable if we extend the `.treetop` file.

I agree with supporting nothing earlier than Ruby 2.5. I wrote this gem a long time ago and didn't get around to updating its dependencies. I'm going to try to pull from your repo, see if there's anything else needing updates, and then push to master with all dependencies updated.

**Comment by jarrett** [comment]
What's the rationale for the commit removing sanitization (https://github.com/Dahie/rbbcode/commit/d19b45019a3c0cf4d342243f22efc92a5e405fe9)?

**Comment by jarrett** [comment]
By the way, when I updated Treetop to the latest version (1.6.10), I started getting the following error upon compiling the `.treetop` file:

```
NoMethodError: undefined method `expected' for #<Treetop::Runtime::SyntaxNode:0x0000557d8de861c8>
    /usr/local/bundle/gems/treetop-1.6.9/lib/treetop/compiler/node_classes/parenthesized_expression.rb:9:in `expected'
```

After a quick look at `parenthesized_expression.rb` in the Treetop source, I tend to think this is a Treetop bug. Reverting Treetop to 1.5.3 fixed the problem. (I also tried all later versions, and all raise the same error.) So I'll keep Treetop pinned to 1.5.3 for now. Let me know if you have any further insight into this.

Beyond that, I updated all dependencies and pushed to master.

**Comment by Dahie** [comment]
I ran into the same issue trying to update treetop and decided to stick with the old version. Good enough.

**Comment by Dahie** [comment]
> What's the rationale for the commit removing sanitization ([Dahie@d19b450](https://github.com/Dahie/rbbcode/commit/d19b45019a3c0cf4d342243f22efc92a5e405fe9))?

You can ignore the other branches in my repo. They were some experiments, but not for merging.

**Comment by jarrett** [comment]
Ok great, thanks! I'd suggest you pull from my master branch when you have
a chance, since that'll get you all the updated dependencies. It shouldn't
impact your new-feature development either way, though.

I'll give some thought to the best way to extend the .treetop file.

**Comment by jarrett** [comment]
This is taking me longer than I expected, but I have a partly working implementation on the [master branch](https://github.com/jarrett/rbbcode/). I pulled from your repo, did some refactoring, and made an attempt at solving the blockquote problem. But there are still some bugs with paragraphs and line breaks. I'll keep working on those.

**Comment by jarrett** [comment]
I think I've worked out the paragraph and line-break bugs. Let me know what you think. We may be close to releasing a new version with Markdown support.

**Comment by Dahie** [comment]
Thank you looks very good! Only, the readme does not mention the markdown output feature and how to use it.

Feel free to close my PR when you release the new version :)

**Comment by jarrett** [comment]
The new version (1.1.0) is released. Thanks again for your help!

**Pull request closed without merging**